cpu/gd32v/periph/gpio.c

```diff
@@ -0,0 +1,243 @@
+}
+
+void gpio_init_analog(gpio_t pin)
+{
+    /* enable the GPIO port RCC */
+    _port_enable_clock(pin);
+
+    /* map the pin as analog input */
+    int pin_num = _pin_num(pin);
+    *(uint32_t *)(&_port(pin)->CTL0 + (pin_num >= 8)) &= ~(0xfl << (4 * (pin_num - ((pin_num >= 8) * 8))));
+}
+
+int gpio_read(gpio_t pin)
+{
+    GPIO_Type *port = _port(pin);
+    unsigned pin_num = _pin_num(pin);
+
+    if (_pin_is_output(port, pin)) {
+        /* pin is output */
+        return (port->OCTL & (1 << pin_num));
+    }
+    else {
+        /* or input */
+        return (port->ISTAT & (1 << pin_num));
+    }
+}
+
+void gpio_set(gpio_t pin)
+{
+    _port(pin)->BOP = (1 << _pin_num(pin));
+}
+
+void gpio_clear(gpio_t pin)
+{
+    _port(pin)->BC = (1 << _pin_num(pin));
+}
+
+void gpio_toggle(gpio_t pin)
+{
+    if (gpio_read(pin)) {
+        gpio_clear(pin);
+    }
+    else {
+        gpio_set(pin);
+    }
+}
+
+void gpio_write(gpio_t pin, int value)
+{
+    if (value) {
+        gpio_set(pin);
+    }
+    else {
+        gpio_clear(pin);
+    }
+}
+
+#ifdef MODULE_PERIPH_GPIO_IRQ
+int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+                  gpio_cb_t cb, void *arg)
+{
+    int pin_num = _pin_num(pin);
+
+    /* disable interrupts on the channel we want to edit (just in case) */
+    EXTI->IMR &= ~(1 << pin_num);
+    /* configure pin as input */
+    gpio_init(pin, mode);
+    /* set callback */
+    exti_ctx[pin_num].cb = cb;
+    exti_ctx[pin_num].arg = arg;
+    /* enable alternate function clock for the GPIO module */
+    periph_clk_en(APB2, RCC_APB2ENR_AFIOEN);
+    /* configure the EXTI channel */
+    AFIO->EXTICR[pin_num >> 2] &= ~(0xf << ((pin_num & 0x3) * 4));
+    AFIO->EXTICR[pin_num >> 2] |=  (_port_num(pin) << ((pin_num & 0x3) * 4));
+    /* configure the active flank */
+    EXTI->RTSR &= ~(1 << pin_num);
+    EXTI->RTSR |=  ((flank & 0x1) << pin_num);
+    EXTI->FTSR &= ~(1 << pin_num);
+    EXTI->FTSR |=  ((flank >> 1) << pin_num);
+    /* active global interrupt for the selected port */
+    if (pin_num < 5) {
+        NVIC_EnableIRQ(EXTI0_IRQn + pin_num);
+    }
+    else if (pin_num < 10) {
+        NVIC_EnableIRQ(EXTI9_5_IRQn);
+    }
+    else {
+        NVIC_EnableIRQ(EXTI15_10_IRQn);
+    }
+    /* clear event mask */
+    EXTI->EMR &= ~(1 << pin_num);
+    /* unmask the pins interrupt channel */
+    EXTI->IMR |= (1 << pin_num);
+    return 0;
+}
+
+void gpio_irq_enable(gpio_t pin)
+{
+    EXTI->IMR |= (1 << _pin_num(pin));
+}
+
+void gpio_irq_disable(gpio_t pin)
+{
+    EXTI->IMR &= ~(1 << _pin_num(pin));
+}
+
+void isr_exti(void)
+{
+    /* only generate interrupts against lines which have their IMR set */
+    uint32_t pending_isr = (EXTI->PR & EXTI->IMR);
+    for (unsigned i = 0; i < GPIO_ISR_CHAN_NUMOF; i++) {
+        if (pending_isr & (1 << i)) {
+            EXTI->PR = (1 << i);        /* clear by writing a 1 */
+            exti_ctx[i].cb(exti_ctx[i].arg);
+        }
+    }
```

Comment:
If you keep it like this, I bet @benpicco will create a follow up PR to covert this to [`bitarithm_test_and_clear()`](https://api.riot-os.org/bitarithm_8h.html#a90bfe935dcde09afd2918a3768ad2b17) the second this gets merged ;-D